model/src/main/scala/hmda/model/institution/Institution.scala

```diff
@@ -0,0 +1,53 @@
+package hmda.model.institution
+
+/**
+ * A financial institution, geared towards requirements for filing HMDA data.
+ */
+case class Institution(
+    id: Int,
+    name: String,
+    externalIds: Set[ExternalId],
+    agency: Agency,
+    institutionType: InstitutionType
+) {
+
+  val extIdsByType: Map[ExternalIdType, ExternalId] = externalIds.map(extId => (extId.idType, extId)).toMap
+
+  /**
+   * Derives the respondentId for a given Institution based on [[hmda.model.institution.Agency]] and [[hmda.model.institution.InstitutionType]],
+   * the rules for which can be found in section "1.4 - Respondent Identification Numbers for 2017 HMDA Filers" of the
+   * <a href="http://www.consumerfinance.gov/data-research/hmda/static/for-filers/2017/2017-HMDA-File-Specifications.pdf">2017 HMDA File Specifications</a>
+   */
+  def respondentId: Either[InvalidRespondentId, ExternalId] = {
+
+    institutionType.depositoryType match {
+      case None => Left(NoDepositoryTypeForInstitutionType(id, institutionType))
+      case Some(dt) =>
+
+        agency.externalIdTypes.get(dt) match {
+          case None => Left(UnsupportedDepositoryTypeByAgency(id, agency, dt))
+          case Some(extIdType) =>
+
+            extIdsByType.get(extIdType) match {
+              case None => Left(RequiredExternalIdNotPresent(id, extIdType))
+              case Some(extId) => Right(ExternalId(extId.id, extIdType))
+            }
+        }
+    }
```

Comment:
I agree, I don't love the three-deep `match` logic.  That said, I personally find it more readable than the idiomatic alternative.  However, if others prefer the above style, I'm not opposed.

I do think this should have real error handling around it, not just wrapped in an `Option`.  There are multiple scenarios where data could be misaligned, and having a good error about _how_ the data is wrong could be key to fixing it.